package.json

```diff
@@ -23,7 +23,8 @@
     "lint": "lerna run --stream lint",
     "test": "lerna run --stream test",
     "start": "lerna run --stream start",
-    "clean": "lerna run --stream clean"
+    "clean": "lerna run --stream clean",
+    "publish-npm": "lerna publish --conventional-commits --create-release github --yes"
```

Comment:
This is equivalent to running `semantic-release`. `--conventional-commits` generates the changelog, and uses the conventional spec to determine which bump to apply (the usual `feat`/`fix`/`BREAKING CHANGE` stuff). `--create-release github` publishes a GitHub release with the changelog. `--yes` just skips the confirmation prompts.

Under the hood, lerna uses [`lerna changed`](https://github.com/lerna/lerna/tree/master/commands/changed#readme) to figure out which packages to publish versions for, it then uses [`lerna version`](https://github.com/lerna/lerna/tree/master/commands/version#readme) to figure out which version to publish. My understanding is that packages that don't update for a given release will stay on an older version, while packages that have had an update will all jump to the same version.

Conceptually this makes sense to me, since we'll want to be able to point consumers to a unified version of each package.

Options and alternatives:

We can use [GitHub actions `schedule` feature](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#schedule) to preform weekly/daily/whatever releases as a means of adding predictability to the version bump schedule.

Another option would be adding a `release` branch that we merge into when we're ready to make a release.

We could also take the approach Carbon has with `lerna publish from-package` and simply merge a PR with updated `package.json` versions whenever a release is needed. The automation would take over from that point and generate the appropriate changelogs and releases. There are some risks with this approach (additional manual work) but it would give each package better control over it's version.

We can also combine a few of these approaches - Weekly automated conventional-commits releases with intermediate from-package releases as needed.

Reply:
Awesome, that makes sense and is very helpful. I guess the question then just becomes, "when and how do we run `npm-publish`"?

> Another option would be adding a release branch that we merge into when we're ready to make a release.

I think this makes the most sense to me. So far we've followed the _"what's in `master` is what's in production (`latest` version)"_ mantra by relying on Travis CI to run the build/publish step when commits are merged into master. This also democratizes releases by allowing contributors to merge their own (approved) PRs and the CI contain the the publish process and github/npm access tokens as secure env variables.

What if the release branch was `master`? With semantic-release we were going to achieve this and follow a time based release schedule by redirecting all new work to a new `next` branch, which would publish on every PR to the `next` channel. On timed intervals (2 weeks, 1 month, whatever), we would PR `next` into `master`, which would trigger a new release on `latest`. Caveat there being that while the `next -> master` PR is open, no new work can go into `next` - all PRs would be temporarily halted until it's merged. Consumers would have the option of following the `next` releases (the same experience today) or follow the time based `latest`. [Semantic release's docs do a better job of explaining](https://github.com/semantic-release/semantic-release/blob/311c46580f57fbe809735d61fea498789d32a61e/docs/recipes/distribution-channels.md#publishing-on-distribution-channels) 😅 

Additionally, if we needed to patch an old version we could open a new `2.100.x` branch that would publish commits on that branch only as patches, `2.100.1`, `2.100.2`, etc

By no means is this how I think we should do it moving forward, more of just an assessment of what @davidicus and I have been discussing for a few months and how we've relied on semantic-release. This boils down to a few open questions:

* How do we support rapid iteration with product teams via (ideally automated) `next` and `latest` distribution channels?
* How would we do maintenance releases to patch old, non-current minor versions without running the publish manually? (rare problem, but could happen)
* Could we augment the publish steps to include posting a comment to PRs with what `next` and `latest` version it was published in? (this has been super helpful in tracking down the source of bugs)
* Would it be better or easier to contain all this automation in Travis, Circle CI (standardize on what Carbon's using) or GH Actions?

Reply:
I've opened PR https://github.com/carbon-design-system/carbon-addons-iot-react/pull/1694 to begin to address some of this. I've added some base workflows for CI/test actions and one to test doing our publish step from an action. These are so much faster than Travis CI.

I've answered a lot of my own questions:

> - How do we support rapid iteration with product teams via (ideally automated) next and latest distribution channels?

Actions! It's clearer to me now how we can tie specific actions to specific branches and run scripts (semantic-release, lerna, etc) with params to set up automated publishing for `next`/`latest`.

> - How would we do maintenance releases to patch old, non-current minor versions without running the publish manually? (rare problem, but could happen)

Set up an action for a branch that is created from the tag we want to start from. Create 2.100.x branch at tag 2.100.0 - new PRs to this branch should only increment patch. Have the action auto generate these new tags as appropriate and publish to npm. 

Actions can also be ran automatically on push/pull request/etc, but also can be triggered manually. Carbon did this recently with their storybook deploy - https://github.com/carbon-design-system/carbon/pull/7061

> - Could we augment the publish steps to include posting a comment to PRs with what next and latest version it was published in? (this has been super helpful in tracking down the source of bugs)

Yes! We can upload artifacts from different github action jobs and download them into new jobs. The publish command could output the version to be released, and on successful release post a comment to the issue with the issue number via [actions/github-script](https://github.com/actions/github-script#comment-on-an-issue)

> - Would it be better or easier to contain all this automation in Travis, Circle CI (standardize on what Carbon's using) or GH Actions?

I think GH Actions is a great place to start.

Reply:
Oh wow somehow I missed all this! Totally agree with all the above though!